### PR TITLE
Support Deepstack in qwen3-omni

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

Original work #2729
* rebase and minor fixes
* add unit test TestDeepstackProcess

# Tests
qwen3_omni_layer_test.py pass locally
The pylint error is unrelated and will be fixed by #3219  

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/decoders.py [hunk 1]
```diff
@@ -848,6 +873,14 @@ def __call__(
                 "nope_layer_interval": self.config.nope_layer_interval,
                 "interleave_moe_layer_step": self.config.interleave_moe_layer_step,
             }
+          # Deepstack visual embedding injection is not supported with scan_layers=True
+          # Use scan_layers=False to enable deepstack
+          if deepstack_visual_embeds is not None and image_masks is not None:
+            raise NotImplementedError(
+                "Deepstack visual embedding injection requires scan_layers=False. "
+                "Set scan_layers=False in your config to use deepstack features."
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/layers/decoders.py]: I move the checking of deepstack_visual_embeds and scan_layers to types.py, this block in decoders.py is removed

File: src/maxtext/layers/decoders.py [hunk 2]
```diff
@@ -928,6 +961,12 @@ def __call__(
             if kv_caches is not None and kv_cache is not None:
               kv_caches[lyr] = kv_cache
 
+            if deepstack_visual_embeds is not None and lyr < len(deepstack_visual_embeds):
+              visual_embeds = deepstack_visual_embeds[lyr]
+              # Use image_masks to identify visual token positions
+              if bidirectional_mask is not None and visual_embeds is not None:
+                y = self._deepstack_process(y, bidirectional_mask, visual_embeds)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/layers/decoders.py]: done

File: src/maxtext/layers/decoders.py [hunk 3]
```diff
@@ -691,6 +691,30 @@ def apply_output_head(self, shared_embedding: nn.Module | nnx.Module, y, determi
...
+
+    Args:
+      hidden_states: [batch, seq_len, hidden_dim] decoder hidden states
+      bidirectional_mask: [batch, seq_len] boolean mask marking visual token positions
+      visual_embeds: [batch, num_visual_tokens, hidden_dim] visual features from encoder layer
+
+    Returns:
+      Updated hidden_states with visual features added at visual positions
+    """
+    # Expand mask to [batch, seq_len, 1] for broadcasting
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/layers/decoders.py]: I moved _deepstack_process outside the class to a module level function deepstack_process and update the test accordingly. It's cleaner now

File: tests/unit/qwen3_omni_layers_test.py [hunk 4]
```diff
@@ -579,6 +580,88 @@ def test_vision_encoder_single_image(self):
...
+class TestDeepstackProcess(unittest.TestCase):
+  """Tests for Decoder._deepstack_process.
+
+  The method adds deepstack visual embeddings into decoder hidden states at the
+  positions indicated by the bidirectional mask (visual token positions).
+  It uses no instance state, so we call it unbound with None as self.
+  """
+
+  def _deepstack_process(self, hidden_states, bidirectional_mask, visual_embeds):
+    return Decoder._deepstack_process(None, hidden_states, bidirectional_mask, visual_embeds)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tests/unit/qwen3_omni_layers_test.py]: I moved _deepstack_process outside the class to a module level function deepstack_process and update the test accordingly. It's cleaner now

File: src/maxtext/layers/decoders.py [hunk 5]
```diff
@@ -264,6 +264,31 @@ def __call__(
...
+  mask_expanded = bidirectional_mask[:, :, jnp.newaxis]
+  # Use cumsum to map each True position in mask to its index in visual_embeds
+  visual_token_idx = jnp.cumsum(bidirectional_mask, axis=1) - 1  # [batch, seq_len], 0-indexed
+
+  # Gather visual tokens: for each position, get the corresponding visual token
+  batch_idx = jnp.arange(hidden_states.shape[0])[:, jnp.newaxis]  # [batch, 1]
+  visual_embeds_scattered = visual_embeds[batch_idx, visual_token_idx, :]  # [batch, seq_len, hidden]
+
+  # Only add where mask is True: hidden_states += visual_embeds * mask
+  hidden_states = hidden_states + visual_embeds_scattered * mask_expanded
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/layers/decoders.py:288]: In the case of visual_embeds_scattered containing NaN or Inf, we want to error out instead of handling it silently.

File: tests/unit/qwen3_omni_layers_test.py [hunk 6]
```diff
@@ -579,6 +580,84 @@ def test_vision_encoder_single_image(self):
...
+  def test_hidden_states_unchanged_without_visual_tokens(self):
+    """When mask is all-False, hidden states should be returned unchanged."""
+    batch, seq_len, hidden_dim = 2, 6, 4
+    np.random.seed(1)
+    hidden_np = np.random.randn(batch, seq_len, hidden_dim).astype(np.float32)
+    mask = jnp.zeros((batch, seq_len), dtype=bool)
+    visual_embeds = jnp.ones((batch, 1, hidden_dim))
+
+    result = deepstack_process(jnp.array(hidden_np), mask, visual_embeds)
+    np.testing.assert_allclose(np.array(result), hidden_np, rtol=1e-6, atol=1e-6)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tests/unit/qwen3_omni_layers_test.py:660]: Similar to the point above: In the case of visual_embeds_scattered containing NaN or Inf, we want to error out instead of handling it silently. So we don't want to test nan tolerance in this case.

**Review by hengtaoguo — APPROVED** [review]
LGTM, thank you Aireen, let's get this merged!
